src/main/java/com/github/maxopoly/finale/combat/CPSHandler.java

```diff
@@ -0,0 +1,39 @@
+package com.github.maxopoly.finale.combat;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.bukkit.entity.Player;
+
+public class CPSHandler {
+
+	private Map<UUID, List<Long>> playerClicks = new ConcurrentHashMap<>();
+
+	public int getCPS(UUID uuid) {
+		 final long time = System.currentTimeMillis();
+		 	List<Long> clicks = this.playerClicks.get(uuid);
+		 	if (clicks == null) {
+		 		return 0;
+		 	}
+	        final Iterator<Long> iterator = clicks.iterator();
+	        while (iterator.hasNext()) {
+	            if (iterator.next() + 1000L < time) {
```

Comment:
Should be configurable/higher, what if CPS < 3?

Reply:
I don't follow.

Reply:
You need a bigger time window to properly determine average CPS for lower limits. Your implementation heavily favors short bursts and stops working entirely for CPS < 1. It'd be more sensible to determine a minimum time between clicks via `1 / cpsLimit` and then enforcing this time limit inbetween clicks.

Additionally this needs more changes if it's to replace the vanilla handling, for example you can definitely hurt non-living entities (boats, carts etc.)

Reply:
Also this only enforces CPS checks for when you are hitting an entity. It should also take hitting the air into account for CPS.

Reply:
Oh right, well the CPS limiter is mainly for preventing players from jitterclicking/butterflying/autoclicking at absurd CPS, which it does well.

Reply:
Originally I tried handling hitting into air too, by listening for https://wiki.vg/Protocol#Animation_.28clientbound.29 instead but it resulted in absurd CPS counters whenever clicking over blocks, and inaccurate CPS counters when attacking/interacting with entities. The CPS limiter only needs to exclude high CPS during combat.

Reply:
> resulted in absurd CPS counters whenever clicking over blocks

Shouldn't be the case if you properly filter out anything but left clicking with a sword. 

Again, we should have as little as possible in NMS code, CPS limiter listening should be spigot/bukkit only. I'd recommend using [this](https://github.com/DevotedMC/CivModCore/blob/master/src/main/java/vg/civcraft/mc/civmodcore/util/cooldowns/MilliSecCoolDownHandler.java) with `1000 / cpsLimit` as cool down. Also if your CPS limit doesn't actually enforce a CPS limit, but only how often you can actually damage an entity (which will always be less than twice per second as long as you don't tinker with invulnerability periods) that should be aggressively documented, because with a CPS of 8 this CPS limiter would only really do something if you are hitting 4 or more different people twice each within a single second. Simply increasing the "normal" attack speed would absolutely not work as expected/intended and heavily favor auto clickers.

If you can find a way to properly include attacking without hitting an entity do that, otherwise I'd recommend removing this and going for an `EntityDamageByEntityEvent` handler instead, where you look at the players builtin attack cool down. Optionally use protocollib to modify the attack speed sent to the player while keeping it the same server side, so the indicator disappears for the players, but the effect still applies server side. Or just don't do any of this at that point, because having the same behavior, but removing the indicator is kind of weird. Not sure.